File: Graphs/3459. Length of Longest V-Shaped Diagonal Segment/solution.py
class Solution:
    def lenOfVDiagonal(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])

        direction = {
            (1,-1): (-1,-1),
            (-1,-1): (-1,1),
            (-1,1): (1,1),
            (1,1): (1,-1),
        }

        @lru_cache(None)
        def dfs(r, c, need, didTurn, curDirection):
            """
            need = 2 | 0
            next need = 2 if current is 0
                      = 0 if current is 2
                      = 2-need
            """
            step = 0

            row, col = r+curDirection[0], c+curDirection[1]
            if 0 <= row < m and 0 <= col < n and grid[row][col] == need:
                step = max(step, dfs(row, col, 2-need, didTurn, curDirection) + 1)

            if not didTurn: # turn 90 deg clockwise
                nextDirection = direction[curDirection]
                row, col = r+nextDirection[0], c+nextDirection[1]
                if 0 <= row < m and 0 <= col < n and grid[row][col] == need:
                    step = max(step, dfs(row, col, 2-need, True, nextDirection) + 1)
            return step

        res = 0
        for r in range(m):
            for c in range(n):
                if grid[r][c] == 1:
                    for curDirection in direction:
                        res = max(res, dfs(r, c, 2, False, curDirection)+1) # plus 1 for initial step `1`
        return res

File: Graphs/3459. Length of Longest V-Shaped Diagonal Segment/solution.cpp
using pii = pair<int,int>;

struct pair_hash {
    template <class T1, class T2>
    std::size_t operator () (const std::pair<T1,T2> &p) const {
        auto h1 = std::hash<T1>{}(p.first);
        auto h2 = std::hash<T2>{}(p.second);

        // Mainly for demonstration purposes, i.e. works but is overly simple
        // In the real world, use sth. like boost.hash_combine
        return h1 ^ h2;  
    }
};

class Solution {
private:
    int m, n;
    unordered_map<pii, pii, pair_hash> dirs {
        {{1,-1}, {-1,-1}},
        {{-1,-1}, {-1,1}},
        {{-1,1}, {1,1}},
        {{1,1}, {1,-1}},
    };

    int memo[505][505][3][2][7];

    int dfs(int r, int c, int v, pii direction, bool didTurn, vector<vector<int>>& grid) {

        int m = grid.size(), n = grid[0].size();
        int hashD = 2 * direction.first + direction.second + 3;
        if (memo[r][c][v][didTurn][hashD] > 0) return memo[r][c][v][didTurn][hashD];

        int wanted = 2-v;
        int row = r+direction.first, col = c+direction.second;
        if (0 <= row && row < m && 0 <= col && col < n) {
            if (v == 1 && grid[row][col] == 2) {
                memo[r][c][v][didTurn][hashD] = max(memo[r][c][v][didTurn][hashD], dfs(row, col, grid[row][col], direction, didTurn, grid) + 1);
            } else if ((v == 2 || v == 0) && grid[row][col] == wanted) {
                memo[r][c][v][didTurn][hashD] = max(memo[r][c][v][didTurn][hashD], dfs(row, col, grid[row][col], direction, didTurn, grid) + 1);
            }
        }

        if (!didTurn) {
            pii new_dir = dirs[direction];
            memo[r][c][v][didTurn][hashD] = max(memo[r][c][v][didTurn][hashD], dfs(r, c, v, new_dir, 1, grid));
        }
        return memo[r][c][v][didTurn][hashD];
    }
public:
    int lenOfVDiagonal(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();

        int res = 0;
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (grid[i][j] == 1) {
                    for (auto& it : dirs) {
                        res = max(res, dfs(i, j, grid[i][j], it.first, 0, grid)+1);
                    }
                }
            }
        }
        return res;
    }
};
